### PR TITLE
fix: Add headers and utm params to CA request

### PR DESCRIPTION
**Pull request by rafiu007**

### CODE REVIEW

**Comment by arajkumar** [comment]
@rafiu007 Please remove all unwanted changes including lots of space addition noises.

File: src/server.ts [hunk 1]
```diff
@@ -181,10 +184,24 @@ const fetchVulnerabilities = async (reqData) => {
     const headers = {
         'Content-Type': 'application/json',
         'Authorization': 'Bearer ' + config.api_token,
+        'request_id': uuid.v4(),
```
Comment (arajkumar) [inline src/server.ts:187]: You are generating one request id per batch, actually you should do one per batch.

File: src/server.ts [hunk 2]
```diff
@@ -181,10 +184,24 @@ const fetchVulnerabilities = async (reqData) => {
     const headers = {
         'Content-Type': 'application/json',
         'Authorization': 'Bearer ' + config.api_token,
+        'request_id': uuid.v4(),
     };
+
+    let manifestHash = crypto.createHash("sha1").update(reqData.filepath).digest("hex");
```
Comment (arajkumar) [inline src/server.ts:190]: manifestHash should be generated only once from sendDiagnostics, not on each batch.

File: src/server.ts [hunk 3]
```diff
@@ -181,10 +184,24 @@ const fetchVulnerabilities = async (reqData) => {
...
+        url += `&utm_source=${config.source}`;
+    }
+
     if (config.uuid) {
         headers['uuid'] = config.uuid;
     }
+
+    if (config.user_agent) {
+        headers['User-Agent'] = config.user_agent;
+    }
```
Comment (arajkumar) [inline src/server.ts:203]: As mentioned, don't touch User-Agent, utm_content and utm_source is enough

File: src/server.ts [hunk 4]
```diff
@@ -341,7 +359,7 @@ connection.onCodeAction((params, token): CodeAction[] => {
         let codeAction = codeActionsMap[diagnostic.range.start.line + "|" + diagnostic.range.start.character];
         if (codeAction != null) {
             codeActions.push(codeAction);
-   
+
```
Comment (mathur07) [inline src/server.ts:362]: remove unwanted line

File: src/server.ts [hunk 5]
```diff
@@ -230,11 +247,12 @@ function runPipeline(response, diagnostics, packageAggregator, diagnosticFilePat
 }
 
 /* Slice payload in each chunk size of @batchSize */
-function slicePayload(payload, batchSize, ecosystem): any {
+function slicePayload(payload, batchSize, ecosystem, manifestfilePath): any {
     let reqData = [];
     for (let i = 0; i < payload.length; i += batchSize) {
         reqData.push({
             "ecosystem": ecosystem,
+            "filepath": manifestfilePath,
```
Comment (arajkumar) [inline src/server.ts:255]: Why do you need this?

File: src/server.ts [hunk 6]
```diff
@@ -285,10 +303,10 @@ const sendDiagnostics = async (ecosystem: string, diagnosticFilePath: string, co
     const requestPayload = missedItems.map(d => ({package: d.name.value, version: d.version.value}));
     // Closure which adds response into cache before firing diagnostics.
     const cacheAndRunPipeline = response => {
-      cache.add(response);
-      pipeline(response);
+       cache.add(response);
+       pipeline(response);
```
Comment (arajkumar) [inline src/server.ts:307]: Remove unwanted changes

File: src/server.ts (hunk 4 above)
Comment (arajkumar) [inline src/server.ts:362]: Ditto

**Review by arajkumar — CHANGES_REQUESTED** [review]

File: src/config.ts [hunk 7]
```diff
@@ -16,6 +16,8 @@ class Config
     home_dir:                 string;
     uuid:                     string;
     golang_executable:        string;
+    source:                   string;
+    user_agent:               string;
```
Comment (mathur07) [inline src/config.ts:20]: use utm-source to identify the origin of the request

**Review by mathur07 — CHANGES_REQUESTED** [review]

File: src/server.ts (hunk 3 above)
Comment (rafiu007) [reply src/server.ts]: Done.

File: src/server.ts (hunk 2 above)
Comment (rafiu007) [reply src/server.ts]: done.

File: src/server.ts (hunk 1 above)
Comment (rafiu007) [reply src/server.ts]: Changed.

File: src/server.ts [hunk 8]
```diff
@@ -230,11 +242,13 @@ function runPipeline(response, diagnostics, packageAggregator, diagnosticFilePat
...
 
 /* Slice payload in each chunk size of @batchSize */
-function slicePayload(payload, batchSize, ecosystem): any {
+function slicePayload(payload, batchSize, ecosystem, manifestHash, requestId): any {
     let reqData = [];
     for (let i = 0; i < payload.length; i += batchSize) {
         reqData.push({
             "ecosystem": ecosystem,
+            "manifest_hash": manifestHash,
+            "request_id":requestId,
```
Comment (arajkumar) [inline src/server.ts:251]: Why this is sent as request payload? extra query params shall be passed as argument to `fetchVulnerabilities`

File: src/server.ts [hunk 9]
```diff
@@ -271,6 +285,8 @@ const sendDiagnostics = async (ecosystem: string, diagnosticFilePath: string, co
     const diagnostics = [];
     const totalCount = new TotalCount();
     const start = new Date().getTime();
+    let manifestHash = crypto.createHash("sha1").update(diagnosticFilePath).digest("hex");
+    let requestId = uuid.v4();
```
Comment (arajkumar) [inline src/server.ts:289]:
```suggestion
    const manifestHash = crypto.createHash("sha256").update(diagnosticFilePath).digest("hex");
    const requestId = uuid.v4();

```

File: src/config.ts [hunk 10]
```diff
@@ -28,6 +30,8 @@ class Config
         this.home_dir = process.env[(process.platform == 'win32') ? 'USERPROFILE' : 'HOME'];
         this.uuid = process.env.UUID || "";
         this.golang_executable = process.env.GOLANG_EXECUTABLE || 'go';
+        this.source = process.env.source || "";
+        this.user_agent = process.env.user_agent || "";
```
Comment (arajkumar) [inline src/config.ts:34]: Remove unused user_agent
```suggestion
```

**Review by arajkumar — CHANGES_REQUESTED** [review]

File: src/server.ts [hunk 11]
```diff
@@ -181,10 +184,19 @@ const fetchVulnerabilities = async (reqData) => {
     const headers = {
         'Content-Type': 'application/json',
         'Authorization': 'Bearer ' + config.api_token,
+        'request_id': requestId,
```
Comment (arajkumar) [inline src/server.ts:187]: Use http custom header format.. e.g. `X-Request-Id`?

File: src/server.ts [hunk 12]
```diff
@@ -171,7 +174,7 @@ const getCAmsg = (deps, diagnostics, totalCount): string => {
 const caDefaultMsg = 'Checking for security vulnerabilities ...';
 
 /* Fetch Vulnerabilities by component-analysis batch api-call */
-const fetchVulnerabilities = async (reqData) => {
+const fetchVulnerabilities = async (reqData, manifestHash, requestId) => {
```
Comment (arajkumar) [inline src/server.ts:177]: Use type hints.
```suggestion
const fetchVulnerabilities = async (reqData: any, manifestHash: str, requestId: uuid) => {

```

File: src/config.ts [hunk 13]
```diff
@@ -28,6 +29,7 @@ class Config
         this.home_dir = process.env[(process.platform == 'win32') ? 'USERPROFILE' : 'HOME'];
         this.uuid = process.env.UUID || "";
         this.golang_executable = process.env.GOLANG_EXECUTABLE || 'go';
+        this.source = process.env.source || "";
```
Comment (arajkumar) [inline src/config.ts:32]:
```suggestion
        this.utm_source = process.env.utm_source || "";
```

**Review by arajkumar — CHANGES_REQUESTED** [review]

File: src/config.ts [hunk 14]
```diff
@@ -28,6 +30,8 @@ class Config
         this.home_dir = process.env[(process.platform == 'win32') ? 'USERPROFILE' : 'HOME'];
         this.uuid = process.env.UUID || "";
         this.golang_executable = process.env.GOLANG_EXECUTABLE || 'go';
+        this.utm_source = process.env.UTM_SOURCE || "";
+        this.telemetry_id = process.env.TELEMETRY_ID || "";
```
Comment (arajkumar) [inline src/config.ts:34]:
```suggestion
        this.redhat_anonymous_Id = process.env.REDHAT_ANONYMOUS_ID || "";
```

File: src/config.ts [hunk 15]
```diff
@@ -16,6 +16,8 @@ class Config
     home_dir:                 string;
     uuid:                     string;
     golang_executable:        string;
+    utm_source:               string;
+    telemetry_id:             string;
```
Comment (arajkumar) [inline src/config.ts:20]:
```suggestion
    redhat_anonymous_Id:             string;
```

File: src/server.ts [hunk 16]
```diff
@@ -181,10 +184,23 @@ const fetchVulnerabilities = async (reqData) => {
...
+    if (config.utm_source) {
+        url += `&utm_source=${config.utm_source}`;
+    }
+
     if (config.uuid) {
         headers['uuid'] = config.uuid;
     }
+    
+    if (config.telemetry_id) {
+        headers['X-Telemetry-Id'] = config.telemetry_id;
```
Comment (arajkumar) [inline src/server.ts:201]:
```suggestion
        headers['X-Redhat_Anonymous_Id'] = config.redhat_anonymous_Id;

```

**Review by arajkumar — CHANGES_REQUESTED** [review]
Looks good with few nits.

**Review by arajkumar — APPROVED** [review]
> const manifestHash = crypto.createHash("sha256").update(diagnosticFilePath).digest("hex");

Probably we have an issue when linking the CA with SA, because `diagnosticFilePath` will have `file://` and SA won't the same prefix. Let's fix when we have an usecase to link CA and SA.

**Review by mathur07 — APPROVED** [review]

**Pull request merged**